src/lsm/forest_fuzz.zig

```diff
@@ -0,0 +1,390 @@
+
+    fn forest_open_callback(forest: *Forest) void {
+        const env = @fieldParentPtr(@This(), "forest", forest);
+        assert(env.state == .superblock_open);
+        env.state = .forest_open;
+    }
+
+    pub fn compact(env: *Environment, op: u64) void {
+        assert(env.state == .forest_open);
+        env.state = .forest_compacting;
+        env.forest.compact(forest_compact_callback, op);
+        while (env.state == .forest_compacting) env.tick();
+    }
+
+    fn forest_compact_callback(forest: *Forest) void {
+        const env = @fieldParentPtr(@This(), "forest", forest);
+        assert(env.state == .forest_compacting);
+        env.state = .forest_open;
+    }
+
+    fn run(storage: *Storage, fuzz_ops: []const FuzzOp) !void {
+        var env: Environment = undefined;
+
+        try env.init(storage);
+        defer env.deinit();
+
+        // Open the superblock then forest.
+        env.open();
+
+        // The forest should behave like a simple key-value data-structure.
+        // We'll compare it to a hash map.
+        var model = std.hash_map.AutoHashMap(u128, Account).init(allocator);
+        defer model.deinit();
+
+        // We need an op number for compact.
+        var op: u64 = 1;
+
+        for (fuzz_ops) |fuzz_op, fuzz_op_index| {
+            log.debug("Running fuzz_ops[{}] == {}", .{ fuzz_op_index, fuzz_op });
+            // Apply fuzz_op to the forest and the model.
+            switch (fuzz_op) {
+                .compact => {
+                    env.compact(op);
+                    op += 1;
+                },
+                .put => |account| {
+                    const groove = &env.forest.grooves.accounts;
+                    groove.put(&account);
+                    model.put(account.id, account) catch unreachable;
+                },
+                .get => |get| {
+                    switch (get) {
+                        .id => |id| {
+                            const groove = &env.forest.grooves.accounts;
+                            const Groove = @TypeOf(groove.*);
+
+                            // Prefetch by id.
+                            const Getter = struct {
+                                finished: bool = false,
+                                prefetch_context: Groove.PrefetchContext = undefined,
+                                fn prefetch_callback(prefetch_context: *Groove.PrefetchContext) void {
+                                    const getter = @fieldParentPtr(@This(), "prefetch_context", prefetch_context);
+                                    getter.finished = true;
+                                }
+                            };
+                            var getter = Getter{};
+                            groove.prefetch_setup(null);
+                            groove.prefetch_enqueue(id);
+                            groove.prefetch(Getter.prefetch_callback, &getter.prefetch_context);
+                            while (!getter.finished) env.tick();
```

Comment:
@sentientwaffle Now that I have prefetching I'm seeing errors where accounts are not found. Is my use of prefetching here correct?